### ISSUE TITLE
Empty buttons remain visible when they should not

### ISSUE THREAD

**Issue by Zasz333** [kstange/MasqueBlizzBars#3]
When I open the spellbook or talents all empty buttons on the interface are visible even when the option "always show buttons" in edit mode is not enabled.

When closing the spellbook or talents all empty buttons are supposed to hide again. But often they remain visible.

The only way to solve this is to reload the UI or log out and back in.

I tested on a toon without masque and yours and didnt get this problem. I checked with StormFX first on his page before realizing it might be your addition to his mod that causes the problem.

**Comment by kstange** [comment]
I'm not able to reproduce this issue and have not seen it before, even though I play with the configuration you described.  A few questions and requests:

* Are you running other addons?
* What Masque skin are you using?
* Does this happen on every character?
* Does this happen every time you open and close spell book or talents, or is there something else you have to do?
* If you can export and share your edit mode layout string I can try to see if that has any bearing on the issue.

**Comment by StormFX** [comment]
I haven't had a chance to test it yet, but also make sure you don't have "Show Empty Buttons" enabled for your action bar(s).

**Comment by Zasz333** [comment]
**_"I haven't had a chance to test it yet, but also make sure you don't have "Show Empty Buttons" enabled for your action bar(s)."_**
Im very sure I dont have "show empty button" enabled. At least on all bars except #1 which doesnt have that option. I mean, I would see the empty buttons all the time if it were enabled.

**_"Are you running other addons that might affect action bars? Feel free to list everything you have installed if you like."_**
No bar addons, just the default bars. I ruled out sneekee mods and advanced interface options by reproducing the bug without them. No other addons affect the bars.

Right now Im testing my whole list. I disabled everything, log out, enable one, log back in and see if the bug occurs. Will report final findings once Im done.

**_"What Masque skin(s) are you using?"_**
Just masque, masque blizzard bars and masque caith.
Im using caith for my (default) action bars, buffs, debuffs and weakaura.

**_"Does this happen to every action bar or just some of them?"_**
All action bars that are enabled. First bar has empty buttons always visible by design. Cant change that. But its all other bars from two to six that have the problem.

**_"Does this happen on every character?"_**
It seems limited to every max level character with masque enabled. Its the only pattern I found so far. My lvl10 bank characters dont really seem to have the issue.

_"Does this happen every time you open and close spell book or talents, or is there something else you have to do?"_
When I encounter the bug it happens at the first time of the session when opening either. Because those two windows are the only ones that force the action bars to display the empty buttons.
When the bug occurs I usually reload my interface and then it takes abit for the bug to occur again.
Thats also how how I reproduced the bug multiple times.

Just now when I took the screenshots the first opening caused the bug to occur. I disabled one specific addon to test, reloaded my interface and now I can open the spell book and talent windows as much as I want and it doesnt cause the bug. I logged out, logged back in, still with that one addon disabled, first thing I do is open the spell book and the bug occurs.

My layout:
0 39 0 0 0 7 7 UIParent -9.0 4.0 -1 ##$'%,&('&)$+$ 0 1 0 7 7 UIParent -233.0 3.0 -1 ##$%%/&('&(#,# 0 2 0 7 7 UIParent 215.0 4.0 -1 ##$%%/&('&(#,# 0 3 1 5 5 UIParent -5.0 -77.0 -1 #$$$%/&('&(#,# 0 4 1 2 0 MultiBarRight -5.0 0.0 -1 #$$$%/&('&(#,# 0 5 0 8 8 UIParent -315.0 3.0 -1 ##$%%/&('&(#,# 0 6 0 8 8 UIParent -240.0 108.0 -1 ##$%%/&('%(#,# 0 7 1 1 7 MultiBar6 0.0 0.0 -1 ##$$%/&('%(#,$ 0 10 0 7 7 UIParent -174.0 113.0 -1 ##$$&('% 0 11 0 7 7 UIParent -248.5 128.0 -1 ##$$&('%,# 0 12 1 6 0 MainMenuBar 0.0 5.0 -1 ##$$&('% 1 -1 0 2 2 UIParent -1.0 -25.0 -1 ##$#%# 2 -1 1 2 2 UIParent 0.0 0.0 -1 ##$#%( 3 0 1 8 7 UIParent -300.0 250.0 -1 $#3# 3 1 1 6 7 UIParent 300.0 250.0 -1 %#3# 3 2 1 3 5 TargetFrame -10.0 0.0 -1 %#&#3# 3 3 1 0 2 CompactRaidFrameManager 0.0 -7.0 -1 '#(#)#-#.#/#1$3# 3 4 1 0 2 CompactRaidFrameManager 0.0 -5.0 -1 ,#-#.#/#0#1#2( 3 5 1 5 5 UIParent 0.0 0.0 -1 &#*$3# 3 6 1 5 5 UIParent 0.0 0.0 -1 -#.#/#4& 3 7 1 4 4 UIParent 0.0 0.0 -1 3# 4 -1 0 7 7 UIParent -0.0 167.0 -1 # 5 -1 0 3 3 UIParent 1062.0 -345.0 -1 # 6 0 1 2 0 MinimapCluster -10.0 -10.0 -1 ##$#%#&.(()( 6 1 1 2 8 BuffFrame -13.0 -15.0 -1 ##$#%#'+(()( 7 -1 0 3 3 UIParent -0.0 -210.5 -1 # 8 -1 0 7 7 UIParent -741.0 34.0 -1 #'$A%$&7 9 -1 0 8 8 UIParent -437.5 142.0 -1 # 10 -1 1 0 0 UIParent 16.0 -116.0 -1 # 11 -1 1 8 8 UIParent -9.0 85.0 -1 # 12 -1 1 2 2 UIParent -110.0 -275.0 -1 #K$# 13 -1 1 8 8 MicroButtonAndBagsBar 0.0 0.0 -1 ##$#%#&) 14 -1 0 8 8 UIParent -9.0 43.9 -1 ##$#%0 15 0 0 3 3 UIParent 2.0 506.5 -1 # 15 1 1 7 1 MainStatusTrackingBarContainer 0.0 0.0 -1 # 16 -1 1 5 5 UIParent 0.0 0.0 -1 #( 17 -1 1 1 1 UIParent 0.0 -100.0 -1 ## 18 -1 1 5 5 UIParent 0.0 0.0 -1 #- 19 -1 1 7 7 UIParent 0.0 0.0 -1 ##

It should look like this:
https://up.picr.de/46041842ly.jpg

Normal layout on all my characters:
https://up.picr.de/46041810pa.jpg

Same character with spellbook open and empty buttons visible:
https://up.picr.de/46041813cf.jpg

Same character after closing the spellbook:
https://up.picr.de/46041811am.jpg

Im very grateful for both your time. I totally love the brainstorming over this. Big thanks :)

**Comment by kstange** [comment]
I was able to reproduce this issue when I include that little addon in my game.

There's not really a huge difference between reloading and initial login for most things, but it's interesting because notifications on login are an exception to this, which explains why this doesn't come back after a reload.  I looked over the built-in interface code around `MainMenuMicroButton_AreAlertsEnabled` and I don't see an immediate problem, but the code uses the shared `HelpTip` system, so my theory is that the game wants to show you another notification using the same popup frame and it gets other parts of the UI into a weird state because it never moves on from trying to show the profession notification, or something else on the micro button bar.

Generally it's not a good practice to override built-in functions.  The better practice is to use `hooksecurefunc` and then when your function runs, undo whatever the game did.  For example, this would immediately hide any popup created by the main menu micro buttons, but it wouldn't acknowledge them so the buttons would still flash.  Like your original code, it would also catch all popups for any micro button, not just profession ones.

``` lua
hooksecurefunc("MainMenuMicroButton_ShowAlert", function() HelpTip:HideAllSystem("MicroButtons") end)
```

Let me know if you find any ways to reproduce without this addon loaded, otherwise I think we have your answer.
